Protocols: Handle `exists(None)` in GFAL Fix #7398

Geogouz opened the pull request:
With the introduced changes, GFAL no longer tries to execute `exists(None)` as it was always leading to an exception. It now returns the exception `RSEOperationNotSupported` which will silently lead to the same prior behavior as it is caught by:

https://github.com/rucio/rucio/blob/5dcf2f958ee80ae335a88b95f83291b1e5b241eb/lib/rucio/rse/rsemanager.py#L274-L280

Moreover, I removed the `return -1` from

https://github.com/rucio/rucio/blob/5dcf2f958ee80ae335a88b95f83291b1e5b241eb/lib/rucio/rse/protocols/gfal.py#L514-L518

because judging from the underlying C function `gfal2_stat(..)`, it is effectively unreachable. Basically, the helper function translates the C GError into a Python exception of type `gfal2.GError` and so, instead of returning any negative integer, it raises the exception.

[Source:](https://github.com/cern-fts/gfal2-python/blob/69466f56110ec87886d1749d0a8d4529d153c30d/src/Gfal2Context.cpp#L156-L165)

```
Stat Gfal2Context::stat_c(const std::string & path)
{
    ScopedGILRelease unlock;
    GError* tmp_err = NULL;
    Stat st;
    const int ret = gfal2_stat(cont->get(), path.c_str(), &st._st, &tmp_err);
    if (ret < 0)
        GErrorWrapper::throwOnError(&tmp_err);
    return st;
}
```

Geogouz committed:
Protocols: Handle `exists(None)` in GFAL #7398

rdimaio commented on lib/rucio/rse/protocols/gfal.py:
```diff
@@ -28,7 +28,7 @@
 try:
     import gfal2  # pylint: disable=import-error
-except:
+except:  # noqa: E722
```
```suggestion
except Exception:
```

This way you can avoid having the noqa, while also avoiding catching possible system exceptions that we might not want to catch: https://docs.astral.sh/ruff/rules/bare-except/

rdimaio commented on lib/rucio/rse/protocols/gfal.py:
```diff
@@ -344,7 +344,12 @@ def exists(self, path):
         """
-        self.logger(logging.DEBUG, 'checking if file exists {}'.format(path))
+
+        self.logger(logging.DEBUG, f'checking if file exists {path}')
```
Avoid f-string logging: https://github.com/rucio/rucio/issues/6826

Geogouz replied:
Ok, makes sense.

Geogouz committed:
Core & Internals: Add is_method_overridden helper to detect subclass method overrides #7398

Geogouz committed:
Protocols: Remove usage of protocol.exists(None) Fix #7398

Removed completely the usage of protocol.exists(None) and placed the fallback to `write` protocol when exists(..) is not overridden.

Geogouz commented on lib/rucio/rse/rsemanager.py:
```diff
@@ -266,18 +266,18 @@ def exists(rse_settings: types.RSESettingsDict, files, domain='wan', scheme=None
-        protocol.exists(None)
-    except NotImplementedError:
+
+    from rucio.rse.protocols.protocol import RSEProtocol  # Placed it here to avoid possible circular imports
```
Placed that here because in protocol we are importing rsemanager itself (which I guess is not the best practice):

https://github.com/rucio/rucio/blob/981f5455551cfab85e8cc1f3fbb89d7e1dfba98e/lib/rucio/rse/protocols/protocol.py#L24-L33
